File: pocket-ec/src/main/java/com/dididi/pocket/ec/main/message/MessageItemDelegate.java
package com.dididi.pocket.ec.main.message;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.Toast;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.dididi.pocket.core.entity.User;
import com.dididi.pocket.core.fakedata.FakeUser;
import com.dididi.pocket.core.ui.dialog.PocketDialog;
import com.dididi.pocket.ec.R;
import com.dididi.pocket.ec.R2;
import com.dididi.pocket.core.ui.item.SearchBarItem;
import com.dididi.pocket.core.entity.Message;
import com.dididi.pocket.core.delegates.bottom.BottomItemDelegate;
import com.dididi.pocket.ec.main.message.adapter.MessageAdapter;
import com.dididi.pocket.ec.main.message.chat.ChatDelegate;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;


/**
 * Created by dididi
 * on 24/07/2018 .
 */

public class MessageItemDelegate extends BottomItemDelegate
        implements BaseQuickAdapter.OnItemChildClickListener {

    @BindView(R2.id.delegate_msg_message_searchBar)
    SearchBarItem mSearchBar = null;
    @BindView(R2.id.delegate_msg_message_list_view)
    RecyclerView mMsgRecyclerView = null;

    private List<Message> mMsgList = new ArrayList<>();
    private LinearLayoutManager layoutManager = null;

    @Override
    public Object setLayout() {
        return R.layout.delegate_msg_message;
    }

    @Override
    public void onBindView(@Nullable Bundle savedInstanceState, View rootView) {
        mSearchBar.setLeftIcon("{faw-plus}");
        initFakeMessage();
        //设置布局方式
        layoutManager = new LinearLayoutManager(getContext(),
                LinearLayoutManager.VERTICAL, false);
        mMsgRecyclerView.setLayoutManager(layoutManager);
        MessageAdapter mAdapter = new MessageAdapter(R.layout.item_message_list, mMsgList);
        mAdapter.setOnItemChildClickListener(this);
        mMsgRecyclerView.setAdapter(mAdapter);
    }

    @Override
    public void onItemChildClick(BaseQuickAdapter adapter, View view, int position) {
        Message message = (Message) adapter.getItem(position);
        if (null == message) {
            throw new RuntimeException("message can not be null");
        }
        //删除消息
        if (view.getId() == R.id.item_message_delete) {
            PocketDialog dialog = PocketDialog.getInstance(getContext());
            dialog.setTitle("删除聊天")
                    .setContent("确认删除这条聊天记录吗？")
                    .isCancelableOnTouchOutside(true)
                    .setConfirmClickListener(v -> {
                        mMsgList.remove(message);
                        adapter.notifyDataSetChanged();
                        dialog.dismiss();
                    })
                    .setCancelClickListener(v -> dialog.dismiss())
                    .show();
        }
        //置顶消息
        if (view.getId() == R.id.item_message_top) {
            if (position > 0 && position < mMsgList.size()) {
                //先移除该条消息
                mMsgList.remove(message);
                //插入一条位置0的消息
                adapter.notifyItemInserted(0);
                mMsgList.add(0, message);
                //移除position+1的消息（即原来的消息）
                adapter.notifyItemRemoved(position + 1);
                if (0 == layoutManager.findFirstVisibleItemPosition()) {
                    mMsgRecyclerView.scrollToPosition(0);
                }
            } else {
                Toast.makeText(getContext(), "消息已置顶", Toast.LENGTH_SHORT).show();
            }
        }
        if (view.getId() == R.id.item_message_main_layout) {
            //点击跳转Chat页面
            getParentDelegate().getSupportDelegate().start(ChatDelegate.getStartChat(message));
        }
    }

    /**
     * 初始化消息列表 不用管它什么意思
     */
    private void initFakeMessage() {
        User sendUser = FakeUser.getUser("1");
        Message msg1 =
                new Message(sendUser.getName() + " 祝您长命百岁",
                        Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("2"),
                        "27/7/2018");
        Message msg2 = new Message(sendUser.getName() + " 祝您万事如意",
                Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("3"),
                "27/7/2018");
        Message msg3 = new Message(sendUser.getName() + " 祝您心想事成",
                Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("4"),
                "27/7/2018");
        Message msg4 = new Message(sendUser.getName() + " 祝您吉祥如意",
                Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("5"),
                "27/7/2018");
        Message msg5 = new Message(sendUser.getName() + " 祝您事事顺心",
                Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("6"),
                "27/7/2018");
        Message msg6 =
                new Message(sendUser.getName() + " 祝您财源滚滚",
                        Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("7"),
                        "27/7/2018");
        Message msg7 = new Message(sendUser.getName() + " 祝您阖家欢乐",
                Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("8"),
                "27/7/2018");
        Message msg8 = new Message(sendUser.getName() + " 祝您六六大顺",
                Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("9"),
                "27/7/2018");
        Message msg9 = new Message(sendUser.getName() + " 祝您福星高照",
                Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("10"),
                "27/7/2018");
        Message msg10 = new Message(sendUser.getName() + " 祝您福如东海",
                Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("11"),
                "27/7/2018");
        Message msg11 = new Message(sendUser.getName() + " 祝您寿比南山",
                Message.TYPE_RECEIVED, sendUser, FakeUser.getUser("12"),
                "27/7/2018");
        mMsgList.add(msg1);
        mMsgList.add(msg2);
        mMsgList.add(msg3);
        mMsgList.add(msg4);
        mMsgList.add(msg5);
        mMsgList.add(msg6);
        mMsgList.add(msg7);
        mMsgList.add(msg8);
        mMsgList.add(msg9);
        mMsgList.add(msg10);
        mMsgList.add(msg11);
    }
}

File: pocket-ec/src/main/java/com/dididi/pocket/ec/main/message/MessageItemDelegate.kt
package com.dididi.pocket.ec.main.message

import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import android.widget.AdapterView
import android.widget.Toast
import com.chad.library.adapter.base.BaseQuickAdapter
import com.dididi.pocket.core.delegates.PocketDelegate
import com.dididi.pocket.core.delegates.bottom.BottomItemDelegate
import com.dididi.pocket.core.im.IIMCallback
import com.dididi.pocket.core.ui.dialog.PocketDialog
import com.dididi.pocket.ec.R
import com.dididi.pocket.ec.main.message.adapter.MessageAdapter
import com.dididi.pocket.ec.main.message.chat.ChatDelegate
import com.dididi.pocket.ec.main.message.chat.model.C2CChatInfo
import com.dididi.pocket.ec.main.message.chat.model.C2CChatManager
import com.dididi.pocket.ec.main.message.model.MessageInfo
import com.dididi.pocket.ec.main.message.model.MessageManager
import com.gyf.immersionbar.ktx.immersionBar
import com.tencent.imsdk.TIMConversationType
import kotlinx.android.synthetic.main.delegate_msg_message.*


/**
 * Created by dididi
 * on 24/07/2018 .
 */

class MessageItemDelegate : BottomItemDelegate(), BaseQuickAdapter.OnItemChildClickListener {

    private var mMsgList = ArrayList<MessageInfo>()
    private var layoutManager: LinearLayoutManager? = null
    private var mAdapter: MessageAdapter? = null

    override fun setLayout(): Any {
        return R.layout.delegate_msg_message
    }

    override fun onBindChildView(savedInstanceState: Bundle?, rootView: View?) {
    }

    override fun onBindView(savedInstanceState: Bundle?, rootView: View?) {

        //设置布局方式
        layoutManager = LinearLayoutManager(context,
                LinearLayoutManager.VERTICAL, false)
        delegate_msg_message_list_view!!.layoutManager = layoutManager
        mAdapter = MessageAdapter(R.layout.item_message_list, mMsgList)
        mAdapter!!.onItemChildClickListener = this
        delegate_msg_message_list_view!!.adapter = mAdapter
        delegate_msg_message_refresh.setColorSchemeColors(ContextCompat.getColor(context!!, R.color.textColorDark))
        delegate_msg_message_refresh.setOnRefreshListener {
            MessageManager.getInstance().loadConversation(object :IIMCallback{
                override fun onSuccess(data: Any?) {
                    delegate_msg_message_refresh.isRefreshing = false
                    mAdapter!!.updateConversations(data as ArrayList<MessageInfo>)
                    Toast.makeText(context, "刷新成功", Toast.LENGTH_SHORT).show()
                }

                override fun onError(module: String, errCode: Int, errMsg: String?) {
                }
            })
        }
        initFakeMessage()
        delegate_msg_message_layout_spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(parent: AdapterView<*>?) {
            }

            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                val users = resources.getStringArray(R.array.spinner_tim_account)
                val info = MessageInfo().apply {
                    peer = users[position]
                    title = peer
                    lastMessageTime = System.currentTimeMillis()
                }
                mMsgList.add(info)
                mAdapter!!.notifyDataSetChanged()
            }
        }
    }

    override fun getTitleBarId() = R.id.delegate_msg_message_toolbar

    override fun initImmersionBar() {
        immersionBar {
            flymeOSStatusBarFontColor(R.color.textColorWhite)
            keyboardEnable(true)
        }
    }

    override fun onItemChildClick(adapter: BaseQuickAdapter<*, *>, view: View, position: Int) {
        val message = adapter.getItem(position) as MessageInfo?
                ?: throw RuntimeException("message can not be null")
        //删除消息
        if (view.id == R.id.item_message_delete) {
            val dialog = PocketDialog.getInstance(context)
            dialog.setTitle("删除聊天")
                    .setContent("确认删除这条聊天记录吗？")
                    .isCancelableOnTouchOutside(true)
                    .setConfirmClickListener {
                        mMsgList.remove(message)
                        adapter.notifyDataSetChanged()
                        dialog.dismiss()
                    }
                    .setCancelClickListener { dialog.dismiss() }
                    .show()
        }
        //置顶消息
        if (view.id == R.id.item_message_top) {
            if (position > 0 && position < mMsgList.size) {
                //先移除该条消息
                mMsgList.remove(message)
                //插入一条位置0的消息
                adapter.notifyItemInserted(0)
                mMsgList.add(0, message)
                //移除position+1的消息（即原来的消息）
                adapter.notifyItemRemoved(position + 1)
                if (0 == layoutManager!!.findFirstVisibleItemPosition()) {
                    delegate_msg_message_list_view!!.scrollToPosition(0)
                }
            } else {
                Toast.makeText(context, "消息已置顶", Toast.LENGTH_SHORT).show()
            }
        }
        if (view.id == R.id.item_message_main_layout) {
            //点击跳转Chat页面
            C2CChatManager.getInstance().setCurrentChatInfo(C2CChatInfo().apply { peer = message.peer;chatName = message.peer;type = TIMConversationType.C2C })
            getParentDelegate<PocketDelegate>().supportDelegate.start(ChatDelegate.getStartChat(message.peer!!))
        }
    }

    /**
     * 重写父类的onScrollToTop()方法，RecyclerView返回顶部
     */
    override fun onScrollToTop() {
        //设置toolbar的内容自动滑出来
        delegate_msg_message_appbar!!.setExpanded(true)
        delegate_msg_message_list_view!!.smoothScrollToPosition(0)
    }

    /**
     * 重写父类的onRefresh()方法，刷新页面
     */
    override fun onRefresh() {

    }

    /**
     * @return 是否位于顶部
     */
    override fun isTop(): Boolean {
        return false
    }

    /**
     * 初始化消息列表 不用管它什么意思
     */
    private fun initFakeMessage() {
        MessageManager.getInstance().loadConversation(object : IIMCallback {
            override fun onSuccess(data: Any?) {
                mAdapter!!.addConversations(data as ArrayList<MessageInfo>)
                Toast.makeText(context, "加载成功", Toast.LENGTH_SHORT).show()
            }

            override fun onError(module: String, errCode: Int, errMsg: String?) {
                Toast.makeText(context, "加载失败", Toast.LENGTH_SHORT).show()
            }
        })
    }

}
